pageserver/src/remote_storage/storage_sync/upload.rs

```diff
@@ -1,521 +1,438 @@
+use std::{fmt::Debug, path::PathBuf};
 
-use tracing::{debug, error, warn};
+use anyhow::Context;
+use futures::stream::{FuturesUnordered, StreamExt};
+use tokio::fs;
+use tracing::{debug, error, trace, warn};
 
 use crate::{
     config::PageServerConf,
+    layered_repository::metadata::metadata_path,
     remote_storage::{
-        storage_sync::{
-            compression, fetch_full_index,
-            index::{RemoteTimeline, TimelineIndexEntry, TimelineIndexEntryInner},
-            sync_queue, SyncKind, SyncTask,
-        },
+        storage_sync::{index::RemoteTimeline, sync_queue, SyncTask},
         RemoteStorage, ZTenantTimelineId,
     },
 };
 
-use super::{compression::ArchiveHeader, NewCheckpoint, RemoteIndex};
+use super::{index::IndexShard, SyncData, TimelineUpload};
+
+/// Serializes and uploads the given shard data to the remote storage.
+pub(super) async fn upload_shard<P, S>(
+    conf: &'static PageServerConf,
+    storage: &S,
+    sync_id: ZTenantTimelineId,
+    shard: IndexShard,
+) -> anyhow::Result<()>
+where
+    P: Debug + Send + Sync + 'static,
+    S: RemoteStorage<StoragePath = P> + Send + Sync + 'static,
+{
+    let shard_bytes =
+        serde_json::to_vec(&shard).context("Failed to serialize shard file into bytes")?;
+    let shard_bytes = tokio::io::BufReader::new(std::io::Cursor::new(shard_bytes));
+
+    let shard_path = metadata_path(conf, sync_id.timeline_id, sync_id.tenant_id)
+        .with_file_name(IndexShard::FILE_NAME)
+        .with_extension(IndexShard::FILE_EXTENSION);
+    let shard_storage_path = storage.storage_path(&shard_path).with_context(|| {
+        format!(
+            "Failed to get the shard storage path for local path '{}'",
+            shard_path.display()
+        )
+    })?;
+
+    storage
+        .upload(shard_bytes, &shard_storage_path, None)
+        .await
+        .with_context(|| {
+            format!(
+                "Failed to upload index shard to the storage path '{:?}'",
+                shard_storage_path
+            )
+        })
+}
 
-/// Attempts to compress and upload given checkpoint files.
-/// No extra checks for overlapping files is made: download takes care of that, ensuring no non-metadata local timeline files are overwritten.
+/// Timeline upload result, with extra data, needed for uploading.
+#[derive(Debug)]
+pub(super) enum UploadedTimeline {
+    /// Upload failed due to some error, the upload task is rescheduled for another retry.
+    FailedAndRescheduled,
+    /// No issues happened during the upload, all task files were put into the remote storage.
+    Successful(SyncData<TimelineUpload>),
+    /// No failures happened during the upload, but some files were removed locally before the upload task completed
+    /// (could happen due to retries, for instance, if GC happens in the interim).
+    /// Such files are considered "not needed" and ignored, but the task's metadata should be discarded and the new one loaded from the local file.
```

Comment:
I think it's not a big issue right now: if the metadata stayed the same, we just reread the same bytes again and submit them into the index.
Otherwise, we will read an updated metadata and use it in the index.

Reply:
I'm still not sure that this won't cause any holes in remote data. And it is a bit scary :)

Reply:
But what kind of holes that could cause, I wonder?
We don't really stick to the exact `disk_consistent_lsn`, rather using it to compare (the one that's bigger is the new one) and the rest of the params are not used in the remote data at all.

In such case, we reread the metadata file only, keeping the rest of the `TimelineUpload` unchanged.

So, nothing is really lost, IMO.

Reply:
I'm thinking about some weird retry combination which can take some time that may be enough for GC to kick in

Reply:
In that case, we would upload actually more than needed, so not really loosing data? 😄 
But sure, GC can do any magic things to our state since we ignore it.